src/OrleansRuntime/Catalog/Catalog.cs

```diff
@@ -263,6 +263,50 @@ public Task CollectActivations(TimeSpan ageLimit)
                 .ToList();
         }
 
+        public List<DetailedGrainStatistic> GetDetailedGrainStatistics()
+        {
+            var checkedTypes = new Dictionary<Type,bool>();
```

Comment:
An alternative approach is to have `GrainTypeManager` build the list of grain types subject to such queries once, and get it here instead of rebuilding it every time. Would also make this method simpler and cleaner.